Add dnn self supervised learning example

arrufat opened the pull request:
Hi, recently I've been interested in self-supervised learning (SSL) methods in deep learning.
I noticed that dlib has support for unsupervised loss layers by just leaving out the `training_label_type` `typedef` and the truth iterator argument to `compute_loss_value_and_gradient()`.

However, most methods I've read about and tried are quite complicated (hard-negative mining, predictors, stop gradient, exponential moving average of the weights between both architectures) due to the need of breaking the symmetry between both branches to avoid collapse. But recently, I came across a simple method named [Barlow Twins: Self-Supervised Learning via Redundancy Reduction](https://arxiv.org/abs/2103.03230). The idea is as follows:

1. forward two augmented versions of an image
1. compute the empirical cross-correlation matrix between both feature representations
1. make that matrix as close to the identity as possible.

That prevents collapse, since it's making each individual dimension of the feature vectors to focus on different tasks, and thus avoiding redundancies between dimensions (it needs a high dimension for it to work, since it relies on sparse representations).

So far, I've implemented:

### `input_rgb_image_pair`

It takes a `std::vector<std::pair<matrix<rgb_pixel>, matrix<rgb_pixel>>>` and puts the first elements of the pairs in the first half of the batch and the second elements of the batch in the second half of the batch. This allows computing batch normalization on each half efficiently (done on the loss layer).

### `loss_barlow_twins`

I tried to follow the official paper as close as possible, and made use of the awesome [Matrix Calculus](http://www.matrixcalculus.org/) site for the gradients (having element wise operations, (off-)diagonal terms, and summations was a bit tedious to do manually 😅)

But… I am experiencing some difficulties:

- if I use a `dnn_trainer` I get a segfault (but if I train manually, like in the code, it works: at least the loss goes down)
- if I use a batch size larger than 64, I also get a segfault (but I have plenty of RAM/VRAM)

So maybe I misunderstood something about how to implement the input layer or the unsupervised loss layer… If you could have a look at some point… You can just run the example by giving a path to a folder containing the CIFAR-10 dataset.

It would be great to have an example of SSL on dlib.

Thanks in advance (all the code is inside the example program, I will make a proper PR if we manage to get this work, and you are interested, since the method is fairly new)

arrufat commented:
I fixed the crash with larger batch sizes (wrong pointer arithmetic).
However, the `dnn_trainer` still crashes after the 3rd iteration... I can't see what I am doing wrong, but I am probably missing something in the unsupervised loss layer...

davisking commented on examples/dnn_self_supervised_learning_ex.cpp:
```diff
@@ -0,0 +1,412 @@
+            alias_tensor split(batch_size, sample_size);
+            alias_tensor sample(1, sample_size);
+            auto za = split(output_tensor);
+            auto zb = split(output_tensor, batch_size);
```
Is this what you want?  That is, offsetting the memory by `batch_size`?  Not some larger size?  Like this is going to overlap `za` isn't it?

arrufat replied:
Ah! right, it has to be `batch_size` * `sample_size`... so that I jump to the second half of the batch...
It still crashes, but that was definitely wrong, thank you!

davisking commented:
I'm not sure.  I looked a bit but I'm not totally sure what's supposed to happen.  I didn't see anything that would obviously lead to a crash though.  You should run it with `ENABLE_ASSERTS` on to see if they catch anything.  And run on the CPU to see if that makes it easier to debug.  You could also run it with valgrind.  Might be super slow though.  But should show you where it's crashing if gdb isn't catching it.  Like with gdb you should be able to get a stack trace showing where the crash is.  Not sure if you tried that yet.  But if that's not doing it somehow valgrind can be useful.

arrufat commented:
Thank you, I will try valgrind :)

davisking commented:
Try `gdb` too.  Not sure if you did that.  But normally gdb will tell you right off where a crash happens and what's going on without any fuss.

arrufat commented:
GDB didn't give me a trace :(
Valgrind did, but I am not sure that I have the skill to understand it yet... If you want to read it, here it is: https://gist.github.com/arrufat/91f7e4862b2567f0be07e2b07b10c580

arrufat commented:
It looks like it's crashing on this line:
https://github.com/davisking/dlib/blob/b9f04fdc453df516f8069f900d456e39f57f3fe1/dlib/dnn/trainer.h#L1214
Did I understand the unsupervised loss interface correctly? I haven't seen any example.

arrufat commented:
Putting that line inside of this condition fixes it:
```c++
if (job.labels[i].size() != 0)
    job.labels[i].assign(lbegin+start, lbegin+stop); 
```
Does that make sense? Since we are not specifying any labels for this unsupervised loss, we can't assign them, right?

arrufat commented:
After running the example for about 2000 steps, this is the empirical cross correlation matrix computed by the loss:
![cross-2048](https://user-images.githubusercontent.com/1671644/134791203-9066d866-20ba-47ce-ab78-72d558192d7e.png)
I think it's working :D

davisking commented:
Yeah, that stack trace is saying there is an illegal memory access at trainer.h:1214, which you have figured out already :)

And yeah, the code at trainer.h:1214 is super wrong :/

pfeatherstone commented:
What a great paper! Nice one @arrufat

pfeatherstone commented:
Wow, the paper used 32 V100 GPUs and trained for 124 hours for 1000 epochs with a batch size of 2048. That's beefy stuff

arrufat commented:
I've refactored the loss and input layers and updated the example.
I've run the example without much tuning of the parameters, and I've got these results
This is the empirical cross-correlation matrix at the beginning of the training:
![eccm-00001](https://user-images.githubusercontent.com/1671644/135749543-4cbe7567-8f7b-46ed-a1f1-ca547b9710bf.png)
And this is what it looks like at the end:
![eccm-04936](https://user-images.githubusercontent.com/1671644/135749577-5151ce48-f866-40bb-b184-067dc03b6247.png)
And this is the training loss:
![image](https://user-images.githubusercontent.com/1671644/135749867-2cf90697-20ee-4300-a84a-4a8814d6a3c2.png)

So, I am quite confident about the implementation. I still need to push more computations to the GPU, though.

On CIFAR-10, I get around 90% accuracy on the train set and around 83% on the test set… I think I should be able to get a bit more. I will try with some different hyperparameters.

Moreover, I have some doubts about the `input_rgb_image_pair` layer, that I want to address before I proceed adding the documentation:

1. Do you think it's really needed?
2. One could achieve the same result with `input_rgb_image` and carefully building each mini-batch (although not that convenient)
3. In case we add `input_rgb_image_pair`, do you think `loss_barlow_twins_` should have a `to_label()` method that re-outputs a `std::vector<std::pair<matrix<rgb_pixel>, matrix<rgb_pixel>>>`?
4. I've also thought about having different average rgb values for the image pair (my first implementation had them) but I don't think I'd ever use it...

arrufat commented:
Ah, I'm assuming you're okay with this PR... if you think it's not a good fit for dlib, there's no problem :)

davisking commented:
Yeah this is super cool.  Totally add it to dlib :)

So my suggestion would be to use `input_rgb_image` and setup the loss like `loss_metric_`.  Since the use of this thing once trained is to get the learned embedding right?  So the training label could be just `bool` and it's telling you which group each training sample is in.  But `output_label_type` would just be `matrix<float,0,1>` since the output, when you are using it, is the embedding vector.  What do you think?

I have mixed feelings about the whole `sample_expansion_factor`.  It seemed like a good idea at the time but in retrospect it's always seemed simpler to use a sample expansion factor of 1 and encode whatever extra information is needed into the labels instead.  Like I had initially imagined that stuff like this and `loss_metric` would use some sample expansion factor > 1.  But the problem is it makes the output of the network, once trained, weird since it keeps forcing the outputs only the loss wants to be the outputs for the actual application, which isn't really what you want.

arrufat commented:
Thank you!
I noticed I made a mistake in my other comment, I meant output `std::vector<std::pair<matrix<float,0,1>, matrix<float,0,1>>>`, so in case we need the embeddings for whatever reason, we can still easily know where they came from.

I totally get what you say about the `sample_expansion_factor`. However, here are some points in favor of `input_rgb_image_pair`:

1. It makes it almost impossible to use it wrong at training time, since it sets up the minibatches for you.
2. With Barlow Twins (and all other SSL methods these days) we have a backbone (such as ResNet50) and a projector (MLP) and after the training is done, we discard the projector and just keep the backbone to finetune on other downstream tasks. So we'll always have to replace the Barlow Twins loss with another loss layer to do something useful with the model.
3. I've made the `input_rgb_image` to be constructible from an `input_rgb_image_pair`, so if we define our models like:

```c++
// This model definition contains the definitions for:
// - SSL model using the Barlow Twins loss, a projector head and an input_rgb_image_pair.
// - Classifier model using the loss_multiclass_log, a fc layer and an input_rgb_image.
namespace model
{
    template <typename SUBNET>
    using projector = fc<128, relu<bn_fc<fc<512, SUBNET>>>>;

    template <typename SUBNET>
    using classifier = fc<10, SUBNET>;

    template <
        template <typename> class LOSS,
        template <typename> class MLP,
        template <typename> class BN,
        typename INPUT>
    using net_type = LOSS<MLP<avg_pool_everything<
        typename resnet::def<BN>::template backbone_50<
        INPUT>>>>;

    using train = net_type<loss_barlow_twins, projector, bn_con, input_rgb_image_pair>;
    using infer = net_type<loss_multiclass_log, classifier, affine, input_rgb_image>;
}
```
We can train our self-supervised model:
```c++
model::train net;
```
And, then assign the subnet in the trained model like:
```c++
model::infer inet;
layer<2>(inet).subnet() = layer<5>(net).subnet();
```

After using it like this, I've found the `input_rgb_image_pair` to be quite useful, but again, it's nothing that couldn't be done when building the mini-batches.

So, in recap, making `input_rgb_image` to be constructible from `input_rgb_image_pair` addresses your concern, right? Since the final embedding learned by the Barlow Twins loss is never going to be used directly.
Again, I am not pushing for `input_rgb_image_pair`, just making sure I understand why `sample_expansion_factor` causes issues.

davisking commented:
Na, I get all that.  It's exactly the setup I added the `sample_expansion_factor` feature for.  I'm just on the fence about it.  Like you could treat this as a supervised problem and have labels that say what bucket each training sample belongs to.  Or do the pair of images per training sample and treat it like it's unsupervised.  I'm legitimately on the fence about which is better.  You know what I mean about setting it up like a supervised problem where the label says which bucket each training sample belongs to?  Like in the Barlow Twins setup each image is in either the A or B bucket.  This could be setup like a supervised training problem where the label of each image says if it's an A or a B.  And the Barlow Twins loss then looks at that label to do what it does.

Anyway, you have been working with it so whatever you find most convenient is best I think.  And like you say, people will almost certainly remove the projector network so they might as well swap out the input layer while they are at it too. And at that point maybe it's less confusing to have explicit input image pairs rather than `A` vs `B` labels :shrug:

arrufat commented:
I am afraid I don't quite follow with the supervised problem setup... The samples not only need to be on the correct bucket, but also in the correct position: the first sample in the first bucket should have it's augmented version in the first sample of the second bucket, so that we can compute the cross-correlation matrix efficiently.
If we do that with `input_rgb_image` the user will have to setup the minibatches like this manually, or the loss won't work. So I don't see how to encode that information in a bool.
And the reason why they samples need to be setup in this way, is to batch normalize both augmented versions of the images independently.
Could you expand on the supervised setup? :)

davisking commented:
Er right, forgot you want cross correlation for this loss. Yeah. You could do it by making the label be an integer id. But yeah, that's more complicated at the user end than having the pair of images setup you have here :)

arrufat commented:
Yes, the think I am worried about, is that if I use the `input_rgb_image` with an integer id, I will have to write a function to build the minibatch properly, and keep copying that function around whenever I want to train the network :P... That is my only concern about using `input_rgb_image`... That's why I added `input_rgb_image_pair`...

davisking commented:
Yeah totally use input_rgb_image_pair :)

arrufat commented:
The other way that I was thinking (still using an unspervised loss, but with `input_rgb_image`) was to build the mini batches like this:
```c++
std::vector<matrix<rgb_pixel>> make_batch(const std::vector<matrix<rgb_pixel>& images, const size_t batch_size)
{
    DLIB_CASSERT(batch_size % 2 == 0);
    const auto offset = batch_size / 2;
    std::vector<matrix<rgb_pixel>> batch(batch_size);
    for (size_t i = 0; i < offset; ++i)
    {
        const auto idx = rnd.get_random_32bit_number() % images.size(); 
        batch[i] = augment(images[idx]);
        batch[i + offset] = augment(images[idx]);
    }
    return batch;
}
```
Which is essentially, what `input_rgb_image_pair` is doing. And then, make sure in the `loss_barlow_twins` that the batch size is even...
That's what I had in mind when I asked if the `input_rgb_image_pair` was needed... It's pretty straighforward, but there's a chance to use it wrong, whereas `input_rgb_image_pair` does it automagically.

Maybe it could even be a static method fo the `loss_barlow_twins_`, so we can call `loss_barlow_twins_::make_batch()`...
I don't know, just putting ideas in the air, so that we might find what's most sensible :)

davisking commented:
Na I think input_rgb_image_pair is the way to go.

pfeatherstone commented:
I wonder if you could combine this loss with loss_yolo to build a network consisting of backbone + detection head + "feature" head such that detection head gave you bounding boxes and class labels, the "feature" head gave you a set of features per bounding box which could be used for tracking, a bit like deep sort.

arrufat commented:
@pfeatherstone maybe, I haven't given it a thought yet... I have a quite selfish approach when adding things to dlib: I only add stuff that I actually use in real-world problems, or that I am really curious about to understand how that works. Unfortunately, I am not that into tracking (yet) so I won't try it any time soon :(

pfeatherstone commented:
@arrufat Absolutely. I think everyone has that mindset. It was just a passing idea I thought I would share.

arrufat commented:
Yes, thank you for sharing your thoughts... honestly, I think I started working on the YOLO loss because you ported the weights first,  so I got really hooked into having the whole thing in dlib... who knows what will happen next?

arrufat commented:
It's ready for review.

I've used a modified ResNet50 (8x downsample, instead of 32x), since the CIFAR-10 images are relatively small (32×32).
According to [this paper](https://arxiv.org/abs/2104.13712), I should be able to get 90% on CIFAR-10 using this method on the test set, but I got 90.798% on the train set and 87.97% on the test set, even though I used an almost identical setup. Maybe learning rate, weight decays and the Adam optimizer work slightly different in PyTorch and dlib... Maybe with some more exhaustive hyperparameter search, I could get the same results, but I think it's good enough for an example program (it takes almost 2 days for the example to complete on an NVIDIA GTX 1080Ti).

That is, assuming I didn't make any mistakes in the loss function or when transferring the pretrained backbone.

Have a look when you have time, and let me know what you think. Having a state-of-the-art self-supervised algorithm in dlib seems pretty cool to me :)

VisionEp1 commented:
Multi gpu training doesnt seem to work anyway even with the small fix. (the white diagonal is missing at the start even).  and the matrix is very gray and collapses eventually 

My guess would be that the inner batches are messed up since  if i renember a batch is build like this:
a1 a2 a3 ... b1 b2 b3 ... and if the trainer splits those without any ordering each net object gets a messed ob vector right?

VisionEp1 commented:
ok good and bad news. 
bad news is im kinda dumb

good news is there is no issue. The training with low batch sizes just collapses and if i increase the gpus i need to increase the batch size since internaly it gets splitted among the gpus. If i increase the batch size * num gpus it seems to work fine

What i would add to the example is that the tensor has the right size regardles of the gpu num, since other dlib examples also can use multiple gpus without issue

davisking commented:
I would define the batch size per GPU. And right, the goal of the example should be mostly to educate, so it's important to make the code simple and tell a single story. Which is a different goal from making a useful tool.  But yeah, setting the batch size per GPU seems like the sensible thing to do in this example.

davisking commented on examples/dnn_self_supervised_learning_ex.cpp:
```diff
@@ -0,0 +1,366 @@
+        serialize("resnet50_self_supervised_cifar_10.net") << layer<5>(net);
+    }
+
+    // To check the quality of the learned feature representations, we will train a linear
+    // classififer on top of the frozen backbone.
+    model::infer inet;
+    // Assign the network, without the projector, which is only used for the self-supervised
+    // training.
+    layer<2>(inet) = layer<5>(net);
+    // Freeze the backbone
+    set_all_learning_rate_multipliers(layer<2>(inet), 0);
```
You use the same augmentation.  You just run the backbone on however many augmented images you want to use as your training set.  Then run the SVM solver on that training set. Like say this training loop here generates 10000 augmented images during training, you could make 10000 augmented images ahead of time and then run the solver on them.  But really you probably need less than there are SGD iterations, since SGD is an awful solver.  The sample efficiency of SGD is really bad.  Anyway, you should try it and see what I mean.

arrufat replied:
I've been re-running the example again with the svm multiclass linear trainer (I deleted the sync files, so I have to run it from scratch again...)
I will try with and without data augmentation and probably make another PR.
Since the CIFAR-10 dataset is quite small, it's feasible to keep all the generated features in memory, but for datasets with millions of images... I don't know... From what I've seen the SVM multiclass linear trainer needs all samples in memory right?

davisking replied:
Right, they all have to be in memory. 

There is only so much capacity in a little linear model like this. At some point it's got all the data and more doesn't change the result. So for really large datasets you likely don't need to load it all anyway.

davisking commented on examples/dnn_self_supervised_learning_ex.cpp:
```diff
@@ -204,7 +204,7 @@ try
     // Initialize the model with the specified projector dimensions and lambda.  According to the
     // second paper, lambda = 1/dims works well on CIFAR-10.
-    model::train net((loss_barlow_twins_(lambda)));
+    model::train net(loss_barlow_twins_(lambda));
```
weird.  Definitely don't need to double up on the () `¯\_(ツ)_/¯`

davisking commented:
Does this build for you?  It's failing on appveyor and when I pull it down and build it with gcc I'm getting the same failure.

davisking commented:
Wow, removing those () is making this error happen.  That is crazy.

davisking commented:
lol, oh right, duh.  Yeah this is the Most Vexing Parse.  `model::train net(loss_barlow_twins_(lambda));` is declaring this function `model::train net(loss_barlow_twins_ lambda)`.  That is, a function called `net` with a single argument named `lambda`

davisking commented:
Anyway, thanks for another cool PR :)

arrufat commented:
Yes, it wasn't building for me without those parenthesis, that's why I added them... I didn't understand why... thank you for the explanation.

arrufat commented:
So, I've tried using the SVM Multiclass linear trainer by adding this block of code:

```c++
{
    console_progress_indicator progress(training_images.size());
    model::feats fnet(layer<5>(net));
    std::vector<matrix<rgb_pixel>> samples;
    std::vector<matrix<float, 0, 1>> features;
    std::vector<unsigned long> labels;
    std::cout << "Generating augmented samples for linear classifier." << std::endl;
    for (size_t i = 0; i < training_images.size(); ++i)
    {
        for (size_t j = 0; j < 10; ++j)
        {
            samples.push_back(augment(training_images[i], false, rnd));
            labels.push_back(training_labels[i]);
        }
        progress.print_status(i, false, std::cerr);
    }
    features = fnet(samples, 2 * batch_size);
    svm_multiclass_linear_trainer<linear_kernel<matrix<float,0,1>>, unsigned long> trainer;
    trainer.set_num_threads(std::thread::hardware_concurrency());
    const auto df = trainer.train(features, labels);
    int num_right = 0, num_wrong = 0;
    for (size_t i = 0; i < features.size(); ++i)
    {
        if (df(features[i]) == labels[i])
            ++num_right;
        else
            ++num_wrong;
    }
    cout << "num right:  " << num_right << endl;
    cout << "num wrong:  " << num_wrong << endl;
    cout << "accuracy:   " << num_right / static_cast<double>(num_right + num_wrong) << endl;
    cout << "error rate: " << num_wrong / static_cast<double>(num_right + num_wrong) << endl;
    features = fnet(testing_images, 2 * batch_size);
    num_right = 0;
    num_wrong = 0;
    for (size_t i = 0; i < features.size(); ++i)
    {
        if (df(features[i]) == testing_labels[i])
            ++num_right;
        else
            ++num_wrong;
    }
    cout << "num right:  " << num_right << endl;
    cout << "num wrong:  " << num_wrong << endl;
    cout << "accuracy:   " << num_right / static_cast<double>(num_right + num_wrong) << endl;
    cout << "error rate: " << num_wrong / static_cast<double>(num_right + num_wrong) << endl;
    return 0;
}
```

But I got accuracies of 80-82% for both training and test sets (as opposed to 88-90% with an fc layer).
Maybe I need to add more data augmentation? Or play a bit with the SVM parameters? What do you think?
I like that the training is really fast, but with the fc layer, the setting seems more flexible, as one could set different learning rates for the backbone and the classifier.

davisking commented:
> So, I've tried using the SVM Multiclass linear trainer by adding this block of code:
> 
> ```c++
> {
>     console_progress_indicator progress(training_images.size());
>     model::feats fnet(layer<5>(net));
>     std::vector<matrix<rgb_pixel>> samples;
>     std::vector<matrix<float, 0, 1>> features;
>     std::vector<unsigned long> labels;
>     std::cout << "Generating augmented samples for linear classifier." << std::endl;
>     for (size_t i = 0; i < training_images.size(); ++i)
>     {
>         for (size_t j = 0; j < 10; ++j)
>         {
>             samples.push_back(augment(training_images[i], false, rnd));
>             labels.push_back(training_labels[i]);
>         }
>         progress.print_status(i, false, std::cerr);
>     }
>     features = fnet(samples, 2 * batch_size);
>     svm_multiclass_linear_trainer<linear_kernel<matrix<float,0,1>>, unsigned long> trainer;
>     trainer.set_num_threads(std::thread::hardware_concurrency());
>     const auto df = trainer.train(features, labels);
>     int num_right = 0, num_wrong = 0;
>     for (size_t i = 0; i < features.size(); ++i)
>     {
>         if (df(features[i]) == labels[i])
>             ++num_right;
>         else
>             ++num_wrong;
>     }
>     cout << "num right:  " << num_right << endl;
>     cout << "num wrong:  " << num_wrong << endl;
>     cout << "accuracy:   " << num_right / static_cast<double>(num_right + num_wrong) << endl;
>     cout << "error rate: " << num_wrong / static_cast<double>(num_right + num_wrong) << endl;
>     features = fnet(testing_images, 2 * batch_size);
>     num_right = 0;
>     num_wrong = 0;
>     for (size_t i = 0; i < features.size(); ++i)
>     {
>         if (df(features[i]) == testing_labels[i])
>             ++num_right;
>         else
>             ++num_wrong;
>     }
>     cout << "num right:  " << num_right << endl;
>     cout << "num wrong:  " << num_wrong << endl;
>     cout << "accuracy:   " << num_right / static_cast<double>(num_right + num_wrong) << endl;
>     cout << "error rate: " << num_wrong / static_cast<double>(num_right + num_wrong) << endl;
>     return 0;
> }
> ```
> 
> But I got accuracies of 80-82% for both training and test sets (as opposed to 88-90% with an fc layer). Maybe I need to add more data augmentation?

No.  If it's low accuracy on the training dataset then adding more augmentation is not the problem.  You are under fitting.  It's totally because you left the C parameter at the default value of 1.  You need to set it to something larger.  It's the same thing as "weight decay".  Small C is like big weight decay.

> I like that the training is really fast, but with the fc layer, the setting seems more flexible, as one could set different learning rates for the backbone and the classifier.

But you froze the backbone.  So the backbone's learning rate and other settings are irrelevant.  And learning rates in general are an SGD hack.  SGD methods ask the user to say "how far should we move each iteration?" because those methods are crappy.  Good solvers will determine the right setting automatically.  Or do something more sophisticated that doesn't even have a "how far did we move?" concept.  Like what this solver is doing.  It's building up a surrogate model that can be solved essentially instantly and perfectly with a special code.  And at every step it knows how much that surrogate deviates from the real learning problem, terminating only when the problem has been provably solved to whatever epsilon you set.

Anyway, think of it like this.  Before computers even existed people knew about gradient descent.  Then entire lifetimes were spent mathematical research coming up with better ways to find the minimums of functions.  There was a lot of progress and wildly faster and more robust methods were developed, this solver is one such method.  The caveat is that these wildly better methods are only applicable to a limited range of problems.  And sadly, none of them is applicable to fitting something like resnet50 :(   They are wildly applicable to fitting linear functions though.

davisking commented:
This is one of the things that drives me up the wall about deep learning.  It's crazy.  There are problems that are solvable in closed form in microseconds to full floating point precision.  But it's easy to find grad students solving them with the SGD golden hammer, and it's taking millions of times longer and they get a solution that's not even close to accurate.

arrufat commented:
First, thank you again for taking your time to give me high-quality feedback, and for bringing me back to reality.
There was a time when, if I had to train a classifier, I would use some kind of SVM or clustering method and carry on with my life.
However, as you said, since deep learning, even the most recent papers by big company names that present state-of-the-art methods for self-supervised learning, perform the linear evaluation by freezing the backbone and training a fully connected layer on top. So, I was following (kind of blindly) what they were doing… Shame on me.

Second, I started playing with the C parameter of the SVM and after a couple of trials (C=5000), I already surpassed the performance of the FC layer with SGD without using any data augmentation: now I am closer to 89% in the test set and over 93% on the training set.

I will prepare a PR later. Thank you again, I will definitely recheck that book. :)

pfeatherstone commented:
Crickey that book is expensive! I miss the good old university days where all of this was either free or at the library.

davisking commented:
>  So, I was following (kind of blindly) what they were doing… Shame on me.

Na, no worries :)   I'm just pointing out the problem with the current resources out there.

davisking commented:
> Crickey that book is expensive! I miss the good old university days where all of this was either free or at the library.

Yeah text books are still expensive.  This one is worth it though.

pfeatherstone commented:
@arrufat you might enjoy this similar paper https://arxiv.org/pdf/2110.07402.pdf it's doing something very similar
